### PR TITLE
Multiple FromParts in handlers

### PR DESCRIPTION
**Pull request by 82marbag**
## Motivation and Context
Handlers are restricted in the number of parameters implementing FromParts; their workaround is to use tuples.

## Description
To aid developers and have better ergonomics, handlers can have up to 8 extensions (FromParts):
```
pub async fn handler(
    input: input::Input,
    ext0: Extension<...>,
    ext1: Extension<...>,
    ext2: Extension<...>,
    ext3: Extension<...>,
    ext4: Extension<...>,
    ext5: Extension<...>,
    ext6: Extension<...>,
    ext7: Extension<...>,
) -> Result<output::Output, error::Error> { ... }
```

## Testing
Tested on PokemonService

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/aws-smithy-http-server/src/rejection.rs [hunk 1]
```diff
@@ -264,21 +264,38 @@ convert_to_request_rejection!(hyper::Error, HttpBody);
 // Required in order to accept Lambda HTTP requests using `Router<lambda_http::Body>`.
 convert_to_request_rejection!(lambda_http::Error, HttpBody);
 
-/// A sum type rejection, implementing [`IntoResponse`] when both variants do.
-pub enum EitherRejection<Left, Right> {
-    Left(Left),
-    Right(Right),
-}
+macro_rules! any_rejection {
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/rejection.rs:267]: Maybe this macro should go down into the `any_rejections` module?

File: rust-runtime/aws-smithy-http-server/src/operation/handler.rs [hunk 2]
```diff
@@ -77,21 +76,36 @@ where
...
-{
-    type Future = Map<Fut, fn(Fut::Output) -> Result<Op::Output, Op::Error>>;
-
-    fn call(&mut self, input: Op::Input, exts: (Ext0, Ext1)) -> Self::Future {
-        (self)(input, exts.0, exts.1).map(IntoResult::into_result)
-    }
+// fn(Input, Ext_i) -> Output
+macro_rules! impl_handler {
+    ($($var:ident),+) => (
+        impl<Op, F, Fut, $($var,)*> Handler<Op, ($($var),*)> for F
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/operation/handler.rs:82]: We'd need to do this for `exts: ($($var,)*))` too

File: rust-runtime/aws-smithy-http-server/src/rejection.rs [hunk 3]
```diff
@@ -264,21 +264,38 @@ convert_to_request_rejection!(hyper::Error, HttpBody);
...
+                match self {
+                    $($name::$var ($var) => $var.into_response(),)*
+                }
+            }
         }
-    }
+    )
+}
+
+pub mod any_rejections {
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/rejection.rs:287]: Should we `doc(hidden)` this?

Comment (LukeMathWalker) [reply rust-runtime/aws-smithy-http-server/src/rejection.rs:287]: The [entire parent module is doc(hidden)](https://github.com/awslabs/smithy-rs/blob/06c23f65cc54e69a6cfdc2526e1e491ba77e5dca/rust-runtime/aws-smithy-http-server/src/lib.rs#L24), so it should not be necessary.

File: CHANGELOG.next.toml [hunk 4]
```diff
@@ -431,3 +431,9 @@ message = "`Endpoint::immutable` now takes `impl AsRef<str>` instead of `Uri`. F
 references = ["smithy-rs#1984", "smithy-rs#1496"]
 meta = { "breaking" = true, "tada" = false, "bug" = false, "target" = "client" }
 author = "jdisanti"
+
+[[smithy-rs]]
+message = "Business logic handlers can have up to 8 extensions in input"
```
Comment (LukeMathWalker) [inline CHANGELOG.next.toml:436]:
```suggestion
message = "Operation handlers can now have up to 8 extensions as input parameters. They were previously limited to 1 extension input parameter."
```

**Review by LukeMathWalker — APPROVED** [review]

File: CHANGELOG.next.toml [hunk 5]
```diff
@@ -431,3 +431,9 @@ message = "`Endpoint::immutable` now takes `impl AsRef<str>` instead of `Uri`. F
 references = ["smithy-rs#1984", "smithy-rs#1496"]
 meta = { "breaking" = true, "tada" = false, "bug" = false, "target" = "client" }
 author = "jdisanti"
+
+[[smithy-rs]]
```
Comment (hlbarber) [inline CHANGELOG.next.toml:435]: This is going to be in https://github.com/awslabs/smithy-rs/pull/1886 changelog? And the handlers used for the old service builder will still be stuck at 1.

Should we remove this entry?

Comment (LukeMathWalker) [reply CHANGELOG.next.toml:435]: Ah, I forgot about that. Yeah, let's roll it up in there. Sorry @82marbag ❤️

**Pull request merged**